counsel.el

```diff
@@ -788,12 +788,26 @@ By default `counsel-bookmark' opens a dired buffer for directories."
       (put-text-property 0 (length key) 'face 'counsel-key-binding key)
       (format "%s (%s)" cmd key))))
 
-(defvar smex-initialized-p)
-(defvar smex-ido-cache)
-(declare-function smex-initialize "ext:smex")
-(declare-function smex-detect-new-commands "ext:smex")
-(declare-function smex-update "ext:smex")
-(declare-function smex-rank "ext:smex")
+
+(defvar smex-initialized-p nil)
+(defvar amx-initialized nil)
```

Comment:
`(defvar var)` and `(defvar var val)` are two entirely different beasts nowadays; the former declares `VAR` as a dynamically bound variable, whereas the latter initialises it as such if it is currently void. We should be doing the former, not initialising external variables. See https://emacs.stackexchange.com/a/41194/15748 for some background on `defvar`.

Reply:
Hi Basil, very thanks for your help. I will -comment your comments-, but they will be really questions in order to understand better how to do this in the future. Your explanation could be very  very clarifying for me.
-------
Thanks for the link here. 
A question: if I declare the variables with ``defvar'' with no initialization value, is it gonna be initialized to nil by default if the symbol is not defined, else how should I use it in a conditional to be save?

Reply:
> if I declare the variables with ``defvar'' with no initialization value, is it gonna be initialized to nil by default if the symbol is not defined

No, a void variable `VAR` will remain void after `(defvar VAR)`, so trying to use its value will signal a `void-variable` error. Calling `defvar` without a second argument only declares the given symbol to the byte-compiler and marks the symbol as "special", i.e. as dynamically, rather than lexically, bound. See <kbd>C-h</kbd><kbd>f</kbd>`defvar`<kbd>RET</kbd>:

    defvar is a special form in `src/eval.c'.

    (defvar SYMBOL &optional INITVALUE DOCSTRING)

    Define SYMBOL as a variable, and return SYMBOL.
    You are not required to define a variable in order to use it, but
    defining it lets you supply an initial value and documentation, which
    can be referred to by the Emacs help facilities and other programming
    tools.  The `defvar' form also declares the variable as "special",
    so that it is always dynamically bound even if `lexical-binding' is t.

    If SYMBOL's value is void and the optional argument INITVALUE is
    provided, INITVALUE is evaluated and the result used to set SYMBOL's
    value.  If SYMBOL is buffer-local, its default value is what is set;
    buffer-local values are not affected.  If INITVALUE is missing,
    SYMBOL's value is not set.

> else how should I use it in a conditional to be save?

To be safe? You check whether its symbol is `boundp` first, before trying to retrieve its `symbol-value`. There is a convenient macro to do this for you, `bound-and-true-p`:

    bound-and-true-p is a Lisp macro in `bindings.el'.

    (bound-and-true-p VAR)

    Return the value of symbol VAR if it is bound, else nil.